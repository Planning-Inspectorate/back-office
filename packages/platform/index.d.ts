import './config/dotenv';
import './http/ttl';
import './util/CryptoUtils';
import './util/date';
import './util/filter';
import './util/url';
import './validators/date';
import './validators/postcode';

declare module '@pins/platform' {
	export * from './config/dotenv';
	export * from './http/ttl';
	export * from './util/CryptoUtils';
	export * from './util/date';
	export * from './util/filter';
	export * from './util/url';
	export * from './validators/date';
	export * from './validators/postcode';
}
