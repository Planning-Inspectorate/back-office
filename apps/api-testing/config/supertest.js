// @ts-nocheck
import supertest from 'supertest';
import config from './config.cjs';
import { Logger } from '../utils/logger.js';

const logger = new Logger();
logger.log('green', '************* Tests starting ****************');
logger.log('cyan', `Running tests against: ${config.baseUrl}`);
logger.log('green', '*********************************************');

const request = supertest(config.baseUrl);

export default request;
