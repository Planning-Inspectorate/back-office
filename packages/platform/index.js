export * from './http/hooks.js';
export * from './testing/assets/assets.js';
export * from './testing/fake.js';
export * from './testing/html-parser.js';
export * from './utils/date.js';
export * from './utils/dotenv.js';
export * from './utils/filter.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
