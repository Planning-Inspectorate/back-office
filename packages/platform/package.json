{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=16.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.0",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.1",
    "@types/lodash-es": "^4.17.6",
    "type-fest": "^2.12.2"
  }
}
