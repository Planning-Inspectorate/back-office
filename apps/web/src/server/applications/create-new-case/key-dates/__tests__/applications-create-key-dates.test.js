import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import { fixtureCases } from '../../../../../../testing/applications/fixtures/cases.js';
import { createTestEnvironment } from '../../../../../../testing/index.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);

const successPatchPostResponse = { id: 123, applicantId: 1 };

const nocks = () => {
	nock('http://test/').patch('/applications/123').reply(200, successPatchPostResponse);
	nock('http://test/')
		.get(/\/applications\/123(.*)/g)
		.reply(200, fixtureCases[0]);
};

describe('Applications create key dates', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	beforeEach(async () => {
		await request.get('/applications-service/');
		nocks();
	});

	afterAll(() => {
		nock.cleanAll();
	});

	const baseUrl = '/applications-service/create-new-case/123/key-dates';

	describe('GET /key-dates', () => {
		it('should render the page', async () => {
			const response = await request.get(baseUrl);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
			expect(element.innerHTML).toContain('Save and continue');
		});
	});

	describe('POST /key-dates', () => {
		describe('Web-side validation:', () => {
			it('should show validation errors if dates are not valid', async () => {
				const response = await request.post(baseUrl).send({
					'keyDates.preApplication.submissionAtInternal.day': 32
				});
				const element = parseHtml(response.text);

				expect(element.innerHTML).toMatchSnapshot();
				expect(element.innerHTML).toContain(
					'Enter a valid day for the anticipated submission date internal'
				);
			});

			it('should show validation errors if submission date is not in the future', async () => {
				const response = await request.post(baseUrl).send({
					'keyDates.preApplication.submissionAtInternal.day': '14',
					'keyDates.preApplication.submissionAtInternal.month': '07',
					'keyDates.preApplication.submissionAtInternal.year': '1789'
				});
				const element = parseHtml(response.text);

				expect(element.innerHTML).toMatchSnapshot();
				expect(element.innerHTML).toContain(
					'The anticipated submission date internal must be in the future'
				);
			});

			it('should redirect to check-your-answers page if there is no error', async () => {
				const response = await request.post(baseUrl).send({
					'keyDates.preApplication.submissionAtInternal.day': '14',
					'keyDates.preApplication.submissionAtInternal.month': '07',
					'keyDates.preApplication.submissionAtInternal.year': '2030'
				});

				expect(response?.headers?.location).toEqual(
					'/applications-service/create-new-case/123/check-your-answers'
				);
			});
		});
	});
});
