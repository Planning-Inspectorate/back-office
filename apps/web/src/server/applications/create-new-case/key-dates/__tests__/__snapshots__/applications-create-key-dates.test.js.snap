// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applications create key dates GET /key-dates should render the page 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <h1 class=\\"govuk-heading-l\\"> Enter the key dates of the project</h1>
    <form method=\\"post\\" action=\\"\\" novalidate=\\"novalidate\\" class=\\"pins-applications-create\\">
        <div class=\\"govuk-form-group\\">
            <label class=\\"govuk-label font-weight--700\\" for=\\"keyDates.submissionDatePublished\\">Anticipated submission date published</label>
            <div id=\\"keyDates.submissionDatePublished-hint\\"
            class=\\"govuk-hint\\">for example, Q4 2023</div>
            <input class=\\"govuk-input govuk-!-width-one-third\\"
            id=\\"keyDates.submissionDatePublished\\" name=\\"keyDates.submissionDatePublished\\"
            type=\\"text\\" aria-describedby=\\"keyDates.submissionDatePublished-hint\\">
        </div>
        <div class=\\"govuk-form-group \\">
            <label class=\\"govuk-label font-weight--700\\" for=\\"submissionDateInternal\\">Anticipated submission date internal</label>
            <div id=\\"submissionDateInternal-hint\\"
            class=\\"govuk-hint\\">for example, 27 3 2023</div>
            <div class=\\"govuk-grid-column-one-third\\">
                <div class=\\"govuk-grid-row pins-date-form-group\\">
                    <div class=\\"govuk-form-group\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalDay\\">Day</label>
                        <input class=\\"govuk-input\\" id=\\"submissionInternalDay\\" name=\\"submissionInternalDay\\"
                        type=\\"text\\">
                    </div>
                    <div class=\\"govuk-form-group\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalMonth\\">Month</label>
                        <input class=\\"govuk-input\\" id=\\"submissionInternalMonth\\" name=\\"submissionInternalMonth\\"
                        type=\\"text\\">
                    </div>
                    <div class=\\"govuk-form-group\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalYear\\">Year</label>
                        <input class=\\"govuk-input\\" id=\\"submissionInternalYear\\" name=\\"submissionInternalYear\\"
                        type=\\"text\\">
                    </div>
                </div>
            </div>
        </div>
        <div class=\\"govuk-grid-row\\">
            <div class=\\"govuk-button-group pins-button-group govuk-grid-column-one-half\\">
                <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Save and continue</button>
            </div>
        </div>
    </form>
</main>"
`;

exports[`Applications create key dates POST /key-dates Web-side validation: should show validation errors if dates are not valid 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <div class=\\"govuk-error-summary\\" aria-labelledby=\\"error-summary-title\\"
            role=\\"alert\\" data-module=\\"govuk-error-summary\\">
                <h2 class=\\"govuk-error-summary__title\\" id=\\"error-summary-title\\"> There is a problem</h2>
                <div class=\\"govuk-error-summary__body\\">
                    <ul class=\\"govuk-list govuk-error-summary__list\\">
                        <li><a href=\\"#keyDates.submissionDateInternal\\">Enter a valid day for the anticipated submission date internal</a>
                        </li>
                    </ul>
                </div>
            </div>
        </div>
    </div>
    <h1 class=\\"govuk-heading-l\\"> Enter the key dates of the project</h1>
    <form method=\\"post\\" action=\\"\\" novalidate=\\"novalidate\\" class=\\"pins-applications-create\\">
        <div class=\\"govuk-form-group\\">
            <label class=\\"govuk-label font-weight--700\\" for=\\"keyDates.submissionDatePublished\\">Anticipated submission date published</label>
            <div id=\\"keyDates.submissionDatePublished-hint\\"
            class=\\"govuk-hint\\">for example, Q4 2023</div>
            <input class=\\"govuk-input govuk-!-width-one-third\\"
            id=\\"keyDates.submissionDatePublished\\" name=\\"keyDates.submissionDatePublished\\"
            type=\\"text\\" aria-describedby=\\"keyDates.submissionDatePublished-hint\\">
        </div>
        <div class=\\"govuk-form-group govuk-form-group--error\\">
            <label class=\\"govuk-label font-weight--700\\" for=\\"submissionDateInternal\\">Anticipated submission date internal</label>
            <div id=\\"submissionDateInternal-hint\\"
            class=\\"govuk-hint\\">for example, 27 3 2023</div>
            <p id=\\"submissionDateInternal-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span> Enter a valid day for
                the anticipated submission date internal</p>
            <div class=\\"govuk-grid-column-one-third\\">
                <div class=\\"govuk-grid-row pins-date-form-group\\">
                    <div class=\\"govuk-form-group govuk-form-group--error\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalDay\\">Day</label>
                        <p id=\\"submissionInternalDay-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span>
                        </p>
                        <input class=\\"govuk-input govuk-input--error\\" id=\\"submissionInternalDay\\"
                        name=\\"submissionInternalDay\\" type=\\"text\\" value=\\"01\\" aria-describedby=\\"submissionInternalDay-error\\">
                    </div>
                    <div class=\\"govuk-form-group govuk-form-group--error\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalMonth\\">Month</label>
                        <p id=\\"submissionInternalMonth-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span>
                        </p>
                        <input class=\\"govuk-input govuk-input--error\\" id=\\"submissionInternalMonth\\"
                        name=\\"submissionInternalMonth\\" type=\\"text\\" value=\\"01\\" aria-describedby=\\"submissionInternalMonth-error\\">
                    </div>
                    <div class=\\"govuk-form-group govuk-form-group--error\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalYear\\">Year</label>
                        <p id=\\"submissionInternalYear-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span>
                        </p>
                        <input class=\\"govuk-input govuk-input--error\\" id=\\"submissionInternalYear\\"
                        name=\\"submissionInternalYear\\" type=\\"text\\" value=\\"2032\\" aria-describedby=\\"submissionInternalYear-error\\">
                    </div>
                </div>
            </div>
        </div>
        <div class=\\"govuk-grid-row\\">
            <div class=\\"govuk-button-group pins-button-group govuk-grid-column-one-half\\">
                <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Save and continue</button>
            </div>
        </div>
    </form>
</main>"
`;

exports[`Applications create key dates POST /key-dates Web-side validation: should show validation errors if submission date is not in the future 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <div class=\\"govuk-error-summary\\" aria-labelledby=\\"error-summary-title\\"
            role=\\"alert\\" data-module=\\"govuk-error-summary\\">
                <h2 class=\\"govuk-error-summary__title\\" id=\\"error-summary-title\\"> There is a problem</h2>
                <div class=\\"govuk-error-summary__body\\">
                    <ul class=\\"govuk-list govuk-error-summary__list\\">
                        <li><a href=\\"#keyDates.submissionDateInternal\\">The anticipated submission date internal must be in the future</a>
                        </li>
                    </ul>
                </div>
            </div>
        </div>
    </div>
    <h1 class=\\"govuk-heading-l\\"> Enter the key dates of the project</h1>
    <form method=\\"post\\" action=\\"\\" novalidate=\\"novalidate\\" class=\\"pins-applications-create\\">
        <div class=\\"govuk-form-group\\">
            <label class=\\"govuk-label font-weight--700\\" for=\\"keyDates.submissionDatePublished\\">Anticipated submission date published</label>
            <div id=\\"keyDates.submissionDatePublished-hint\\"
            class=\\"govuk-hint\\">for example, Q4 2023</div>
            <input class=\\"govuk-input govuk-!-width-one-third\\"
            id=\\"keyDates.submissionDatePublished\\" name=\\"keyDates.submissionDatePublished\\"
            type=\\"text\\" aria-describedby=\\"keyDates.submissionDatePublished-hint\\">
        </div>
        <div class=\\"govuk-form-group govuk-form-group--error\\">
            <label class=\\"govuk-label font-weight--700\\" for=\\"submissionDateInternal\\">Anticipated submission date internal</label>
            <div id=\\"submissionDateInternal-hint\\"
            class=\\"govuk-hint\\">for example, 27 3 2023</div>
            <p id=\\"submissionDateInternal-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span> The anticipated submission
                date internal must be in the future</p>
            <div class=\\"govuk-grid-column-one-third\\">
                <div class=\\"govuk-grid-row pins-date-form-group\\">
                    <div class=\\"govuk-form-group govuk-form-group--error\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalDay\\">Day</label>
                        <p id=\\"submissionInternalDay-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span>
                        </p>
                        <input class=\\"govuk-input govuk-input--error\\" id=\\"submissionInternalDay\\"
                        name=\\"submissionInternalDay\\" type=\\"text\\" aria-describedby=\\"submissionInternalDay-error\\">
                    </div>
                    <div class=\\"govuk-form-group govuk-form-group--error\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalMonth\\">Month</label>
                        <p id=\\"submissionInternalMonth-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span>
                        </p>
                        <input class=\\"govuk-input govuk-input--error\\" id=\\"submissionInternalMonth\\"
                        name=\\"submissionInternalMonth\\" type=\\"text\\" aria-describedby=\\"submissionInternalMonth-error\\">
                    </div>
                    <div class=\\"govuk-form-group govuk-form-group--error\\">
                        <label class=\\"govuk-label govuk-hint\\" for=\\"submissionInternalYear\\">Year</label>
                        <p id=\\"submissionInternalYear-error\\" class=\\"govuk-error-message\\"><span class=\\"govuk-visually-hidden\\">Error:</span>
                        </p>
                        <input class=\\"govuk-input govuk-input--error\\" id=\\"submissionInternalYear\\"
                        name=\\"submissionInternalYear\\" type=\\"text\\" aria-describedby=\\"submissionInternalYear-error\\">
                    </div>
                </div>
            </div>
        </div>
        <div class=\\"govuk-grid-row\\">
            <div class=\\"govuk-button-group pins-button-group govuk-grid-column-one-half\\">
                <button class=\\"govuk-button\\" data-module=\\"govuk-button\\">Save and continue</button>
            </div>
        </div>
    </form>
</main>"
`;
